### PR TITLE
Implemented custom test result export step

### PR DESCRIPTION
**Pull request by Szabadember**
This PR implements the custom test result export step

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+#   name = "github.com/x/y"
+#   version = "2.4.0"
+#
+# [prune]
+#   non-go = false
+#   go-tests = true
+#   unused-packages = true
+
+
+[[constraint]]
```
Comment (godrei) [inline Gopkg.toml:1]: pls remove template code

File: LICENSE [hunk 2]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
+
+Copyright (c) 2020 Áron Szabados
```
Comment (godrei) [inline LICENSE]: Copyright (c) 2020 Bitrise

File: bitrise.yml [hunk 3]
```diff
@@ -0,0 +1,100 @@
+format_version: 4
```
Comment (godrei) [inline bitrise.yml]: pls use the latest format version ([10](https://github.com/bitrise-io/bitrise/blob/master/models/models.go#L23)) to be able to benefit the new version lock (later)

File: bitrise.yml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+default_step_lib_source: https://github.com/bitrise-io/bitrise-steplib.git
+
+app:
+  envs:
+  # An example secret param, define it (A_SECRET_PARAM) in .bitrise.secrets.yml
+  - A_SECRET_PARAM: $A_SECRET_PARAM
```
Comment (godrei) [inline bitrise.yml]: pls remove template code

File: bitrise.yml [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+  audit-this-step:
+    steps:
+    - script:
+        inputs:
+        - content: |-
+            #!/bin/bash
+            set -ex
+            stepman audit --step-yml ./step.yml
+
+  share-this-step:
```
Comment (godrei) [inline bitrise.yml]: we use tooling control center Bitrise project as a centralized release process, so no need to define the release process in the step repo

File: bitrise.yml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+  # An example secret param, define it (A_SECRET_PARAM) in .bitrise.secrets.yml
+  - A_SECRET_PARAM: $A_SECRET_PARAM
+  # If you want to share this step into a StepLib
+  - BITRISE_STEP_ID: custom-test-results-export
+  - BITRISE_STEP_VERSION: "0.0.1"
+  - BITRISE_STEP_GIT_CLONE_URL: https://github.com/bitrise-steplib/bitrise-step-custom-test-results-export.git
+  - MY_STEPLIB_REPO_FORK_GIT_URL: $MY_STEPLIB_REPO_FORK_GIT_URL
+
+workflows:
+  test:
```
Comment (godrei) [inline bitrise.yml]: before_run:
- audit-this-step

File: config.go [hunk 7]
```diff
@@ -0,0 +1,10 @@
+package main
+
+// Config holds the step inputs
+type Config struct {
+	TestName       string `env:"test_name,required"`
+	SearchPath     string `env:"search_path,required"`
+	TestResultsDir string `env:"bitrise_test_result_dir,required"`
+
```
Comment (godrei) [inline config.go]: why the empty line here?

File: config.go [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package main
+
+// Config holds the step inputs
+type Config struct {
+	TestName       string `env:"test_name,required"`
+	SearchPath     string `env:"search_path,required"`
+	TestResultsDir string `env:"bitrise_test_result_dir,required"`
```
Comment (godrei) [inline config.go]: there is a `dir` struct field tag, which also validates if the given dir exists or not

File: main.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	if err := stepconf.Parse(&stepConf); err != nil {
+		failf("Config: %s", err)
+	}
+	stepconf.Print(stepConf)
+
+	log.SetEnableDebugLog(stepConf.VerboseLog)
+
+	matches, err := filepath.Glob(stepConf.SearchPath)
+	if err != nil {
+		failf("Invalid search path %s, error: %s", stepConf.SearchPath, err)
```
Comment (godrei) [inline main.go]: do not use the `, error:` suffix, simple write: `Invalid search path %s: %s`

File: main.go [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	if len(matches) < 1 {
+		failf("Search path did not match any files, path: %s", stepConf.SearchPath)
```
Comment (godrei) [inline main.go]: Search path %s did not match any files

File: main.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	if len(matches) > 1 {
+		log.Warnf("Search path matched more than one file, will use the first match, matches: %s", matches)
```
Comment (godrei) [inline main.go]: default slice String method is not the nicest format.
as Tomi suggested, have something like this:

Search path matched more than one file, will use the first match:
- **match1**
- match2
- match3

highlight the first item with green color.

File: main.go [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+	}
+
+	match := matches[0]
+	exporter := testresultexport.NewExporter(stepConf.TestResultsDir)
+
+	if err := exporter.ExportTest(stepConf.TestName, match); err != nil {
+		failf("Failed to export test result, error: %s", err)
+	}
+
+	os.Exit(0)
```
Comment (godrei) [inline main.go]: no need for this line

File: step.yml [hunk 13]
```diff
@@ -0,0 +1,100 @@
+#
```
Comment (godrei) [inline step.yml]: pls remove template code

File: step.yml [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+#   - android
+#   - xamarin
+#   - react-native
+#   - cordova
+#   - ionic
+
+# Type tags are used for categorizing steps, for easier step discovery in Step Libraries.
+# You can find more information about type tags in the Step Development Guideline:
+# https://github.com/bitrise-io/bitrise/blob/master/_docs/step-development-guideline.md
+type_tags:
```
Comment (godrei) [inline step.yml:13]: for now, we use test tag for steps executing tests, instead, use utility

File: step.yml [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+  - test
+
+is_requires_admin_user: true
+is_always_run: false
```
Comment (godrei) [inline step.yml]: set it to true (if a test fails we still want to export the results)

File: step.yml [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+is_skippable: false
+run_if: ""
+
+deps:
```
Comment (godrei) [inline step.yml]: pls remove unused deps

File: step.yml [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+  - name: wget
+
+
+toolkit:
+  go:
+    package_name: github.com/bitrise-steplib/bitrise-step-custom-test-results-export
+
+
+inputs:
+  - test_name: Custom test result
```
Comment (godrei) [inline step.yml]: pls do not set a default for this input (users will not realize that they need to set it)

File: step.yml [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    opts:
+      title: "The name of the test"
+      summary: Test name displayed on the tab of the Test Reports page.
+      description: |
+        This is the text displayed on the tab of the Test Reports page.
```
Comment (godrei) [inline step.yml]: the description is optional if it is the same, it is better to remove

File: testresultexport/testresultexport.go [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+	destination, err := os.Create(dst)
+	if err != nil {
+		return err
+	}
+	defer destination.Close()
+	_, err = io.Copy(destination, source)
+	return err
+}
+
+func generateTestInfoFile(dir string, data []byte) error {
```
Comment (godrei) [inline testresultexport/testresultexport.go]: fileutil already has this functionality: https://github.com/bitrise-io/go-utils/blob/master/fileutil/fileutil.go

File: testresultexport/testresultexport.go [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+	if err := e.GenerateTestInfoFile(exportDir, testInfoData); err != nil {
+		return err
+	}
+
+	err = e.Copy(testResultPath, testResultExportName)
+
+	return err
+}
+
+func copy(src, dst string) error {
```
Comment (godrei) [inline testresultexport/testresultexport.go]: we already have this functionality: https://github.com/bitrise-io/go-utils/blob/master/command/file.go#L12

File: testresultexport/testresultexport.go [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+
+// ExportTest exports a test result with a given name
+func (e *Exporter) ExportTest(name string, testResultPath string) error {
+	testInfo := &TestInfo{
+		Name: name,
+	}
+
+	testResultName := filepath.Base(testResultPath)
+	exportDir := filepath.Join(e.ExportPath, name)
+	testResultExportName := filepath.Join(exportDir, testResultName)
```
Comment (godrei) [inline testresultexport/testresultexport.go]: extension should be kept

File: testresultexport/testresultexport.go [hunk 22]
```diff
@@ -0,0 +1,123 @@
...
+
+	if err := e.MkdirAll(exportDir, os.ModePerm); err != nil {
+		return fmt.Errorf("skipping test result (%s): could not ensure unique export dir (%s): %s", testResultPath, exportDir, err)
+	}
+
+	if err := e.GenerateTestInfoFile(exportDir, testInfoData); err != nil {
+		return err
+	}
+
+	err = e.Copy(testResultPath, testResultExportName)
```
Comment (godrei) [inline testresultexport/testresultexport.go]: could be: `return e.Copy(testResultPath, testResultExportName)`

**Review by godrei — CHANGES_REQUESTED** [review]
pls check my notes

**Comment by Szabadember** [comment]
test this please

File: bitrise.yml [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+            set -ex
+            stepman audit --step-yml ./step.yml
+  test:
+    before_run:
+    - audit-this-step
+    steps:
+    - go-test:
+        inputs:
+        - packages: |-
+            github.com/bitrise-steplib/step-custom-test-results-export/testresultexport
```
Comment (trapacska) [inline bitrise.yml:20]: we have a legacy: `steps-` and the new `bitrise-step-` prefix for our step repos. please keep on going with the new one.
please dont forget to rewerite everywhaere

File: config.go [hunk 24]
```diff
@@ -0,0 +1,10 @@
+package main
+
+// Config holds the step inputs
+type Config struct {
```
Comment (trapacska) [inline config.go]: pls do not export this as it is unnecessary

File: config.go [hunk 25]
```diff
@@ -0,0 +1,10 @@
...
+	TestName       string `env:"test_name,required"`
+	BasePath       string `env:"base_path,required"`
+	Pattern        string `env:"pattern,required"`
```
Comment (trapacska) [inline config.go]: not really talkative "pattern" what about `search_pattern` or maybe something more talkative. dont forget to modify it in the syep.yml also

File: main.go [hunk 26]
```diff
@@ -0,0 +1,69 @@
...
+
+func failf(format string, args ...interface{}) {
+	log.Errorf(format, args...)
+	os.Exit(1)
+}
+
+func main() {
+	var stepConf Config
+	if err := stepconf.Parse(&stepConf); err != nil {
+		failf("Config: %s", err)
```
Comment (trapacska) [inline main.go]: the error message this package returns matches with the prefix: `Issue with input:`

File: main.go [hunk 27]
```diff
@@ -0,0 +1,69 @@
...
+	}
+	stepconf.Print(stepConf)
+
+	log.SetEnableDebugLog(stepConf.VerboseLog)
+
+	matches := []string{}
```
Comment (trapacska) [inline main.go]: NOTE: `var matches []string`

File: main.go [hunk 28]
```diff
@@ -0,0 +1,69 @@
...
+	basePath := strings.Split(stepConf.BasePath, "*")[0]
+	err := filepath.Walk(basePath, func(path string, info os.FileInfo, err error) error {
```
Comment (trapacska) [inline main.go:32]: NOTE: you can move this into the if

**Review by trapacska — CHANGES_REQUESTED** [review]

File: step.yml [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+
+inputs:
+  - test_name:
+    opts:
+      title: "The name of the test"
+      summary: Test name displayed on the tab of the Test Reports page.
+      is_expand: true
+      is_required: true
+      value_options: []
+  - base_path:
```
Comment (trapacska) [inline step.yml]: please set `$BITRISE_SOURCE_DIR` as default input

**Review by trapacska — CHANGES_REQUESTED** [review]
also pls add check for the output

**Review by trapacska — APPROVED** [review]

**Pull request merged**